Sources/X509/CryptographicMessageSyntax/CMSOperations.swift

```diff
@@ -0,0 +1,222 @@
+            //
+            // > digestAlgorithms is a collection of message digest algorithm
+            // > identifiers.
+            // > ...
+            // > Implementations MAY fail to validate signatures that use a digest
+            // > algorithm that is not included in this set.
+            guard signedData.digestAlgorithms.contains(signer.digestAlgorithm) else {
+                return .init(invalidCMSBlockReason: "Digest algorithm mismatch")
+            }
+
+            // Convert the signature algorithm to confirm we understand it.
+            // We also want to confirm the digest algorithm matches the signature algorithm.
+            let signatureAlgorithm = Certificate.SignatureAlgorithm(algorithmIdentifier: signer.signatureAlgorithm)
+            let expectedDigestAlgorithm = try AlgorithmIdentifier(digestAlgorithmFor: signatureAlgorithm)
+            guard expectedDigestAlgorithm == signer.digestAlgorithm else {
+                return .init(invalidCMSBlockReason: "Digest and signature algorithm mismatch")
+            }
+
+            // Ok, now we need to find the signer. We expect to find them in the list of certificates provided
+            // in the signature.
+            guard let signingCert = try signedData.certificates?.certificate(signerInfo: signer) else {
+                return .init(invalidCMSBlockReason: "Unable to locate signing certificate")
+            }
+
+            // Ok at this point we've done the cheap stuff and we're fairly confident we have the entity who should have
+            // done the signing. Our next step is to confirm that they did in fact sign the data. For that we have to compute
+            // the digest and validate the signature.
+            let signature = try Certificate.Signature(signatureAlgorithm: signatureAlgorithm, signatureBytes: signer.signature)
+            guard signingCert.publicKey.isValidSignature(signature, for: dataBytes, signatureAlgorithm: signatureAlgorithm) else {
+                return .init(invalidCMSBlockReason: "Invalid signature from signing certificate: \(signingCert)")
+            }
+
+            // Ok, the signature was signed by the private key associated with this cert. Now we need to validate the certificate.
+            // This force-unwrap is safe: we know there are certificates because we've located at least one certificate from this set!
+            let untrustedIntermediates = CertificateStore(signedData.certificates!)
+            var verifier = Verifier(rootCertificates: trustRoots, policy: policy)
+            let result = await verifier.validate(leafCertificate: signingCert, intermediates: untrustedIntermediates)
+
+            switch result {
+            case .validCertificate:
+                return .validSignature(.init(signer: signingCert))
+            case .couldNotValidate(let validationFailures):
+                return .unableToValidateSigner(.init(validationFailures: validationFailures))
```

Comment:
Can we also include `signingCert` in `unableToValidateSigner`?

Reply:
Done.

Reply:
🙏